File: experiments/eyecite_us_grammar/performance.py
"""Cold Python reference-path guard for 37,000 US citation inputs."""
from __future__ import annotations

import argparse
import json
import sys
import time
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(ROOT / "legal-pdf-parser" / "src"))

from legalpdf.deterministic_citations import split_footnote_recall_first


CASES = (
    ("Roe v Wade, 410 U.S. 113.", ("reporter",)),
    ("See 410 U.S. at 115.", ("reporter",)),
    ("Claim under 42 U.S.C. § 1983.", ("statute",)),
    ("Article at 123 Harv. L. Rev. 456.", ("journal",)),
)


def checkpoint(path: Path, payload: dict) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    temporary = path.with_suffix(".tmp")
    temporary.write_text(json.dumps(payload, indent=2) + "\n", encoding="utf-8")
    temporary.replace(path)


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--count", type=int, default=37_000)
    parser.add_argument("--max-seconds", type=float, default=15.0)
    args = parser.parse_args()
    output = Path(__file__).parent / "results" / "performance-latest.json"
    started = time.perf_counter()
    interval = max(1, args.count // 10)
    failures: list[dict] = []

    for index in range(args.count):
        text, expected = CASES[index % len(CASES)]
        result = split_footnote_recall_first(text)
        got = tuple(anchor for part in result.parts for anchor in part.anchors)
        if got != expected:
            failures.append({"index": index, "input": text, "expected": expected, "got": got})
        completed = index + 1
        if completed % interval == 0 or completed == args.count:
            elapsed = time.perf_counter() - started
            payload = {
                "format": "beaver.eyecite-us-performance.v1",
                "completed": completed,
                "count": args.count,
                "elapsed_seconds": round(elapsed, 3),
                "failures": failures[:20],
            }
            checkpoint(output, payload)
            print(f"{completed:,}/{args.count:,}; {elapsed:.3f}s", flush=True)

    elapsed = time.perf_counter() - started
    passed = not failures and elapsed <= args.max_seconds
    payload.update({"max_seconds": args.max_seconds, "passed": passed})
    checkpoint(output, payload)
    print(json.dumps(payload, indent=2))
    return 0 if passed else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: experiments/eyecite_us_grammar/performance.ts
/** Cold production-path guard for 37,000 US citation inputs. */
import { mkdirSync, renameSync, writeFileSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

import { citationsInTextNative as citationsInText } from "../../backend/src/lib/structureNative";

const count = 37_000;
const maxSeconds = 15;
const here = path.dirname(fileURLToPath(import.meta.url));
const output = path.join(here, "results", "performance-production-latest.json");
const cases = [
  (index: number) => `Roe v Wade, 410 U.S. ${100 + index}.`,
  (index: number) => `See 410 U.S. at ${100 + index}.`,
  (index: number) => `Claim under 42 U.S.C. § ${1000 + index}.`,
  (index: number) => `Article at 123 Harv. L. Rev. ${100 + index}.`,
];

function checkpoint(payload: object): void {
  mkdirSync(path.dirname(output), { recursive: true });
  const temporary = `${output}.tmp`;
  writeFileSync(temporary, `${JSON.stringify(payload, null, 2)}\n`, "utf8");
  renameSync(temporary, output);
}

const started = performance.now();
const interval = count / 10;
const failures: Array<{ index: number; input: string }> = [];
let payload = {};
for (let index = 0; index < count; index += 1) {
  const input = cases[index % cases.length](index);
  if (citationsInText(input).length === 0 && failures.length < 20) {
    failures.push({ index, input });
  }
  const completed = index + 1;
  if (completed % interval === 0 || completed === count) {
    const elapsedSeconds = (performance.now() - started) / 1000;
    payload = {
      format: "beaver.eyecite-us-production-performance.v1",
      completed,
      count,
      elapsedSeconds: Number(elapsedSeconds.toFixed(3)),
      failures,
    };
    checkpoint(payload);
    console.log(`${completed.toLocaleString()}/${count.toLocaleString()}; ${elapsedSeconds.toFixed(3)}s`);
  }
}

const elapsedSeconds = (performance.now() - started) / 1000;
const passed = failures.length === 0 && elapsedSeconds <= maxSeconds;
payload = { ...payload, maxSeconds, passed };
checkpoint(payload);
console.log(JSON.stringify(payload, null, 2));
process.exitCode = passed ? 0 : 1;
